File: 0278-first-bad-version/0278-first-bad-version.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        let l = 0;
        let r = n;
        
        while(l <= r){
            let mid = l + Math.floor((r-l)/2);
            let test = isBadVersion(mid);
            let prev = isBadVersion(mid-1);
            // console.log(prev);
            // console.log(mid);
            if(test && prev == false){
                return mid;
            } else if(!test){
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        // console.log('def');
        return r;
    };
};

File: 0278-first-bad-version/0278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int minB = n;
        int l = 0;
        int r = n;
        
        while(l <= r){
            int mid = l + (r-l)/2;
            if(isBadVersion(mid)){
                minB = Math.min(mid, minB);
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return minB;
    }
}